data-prepper-plugins/grok-prepper/README.md

```diff
@@ -1,59 +1,241 @@
 
+Creating files of custom patterns makes it easy to organize them. Consider the following directory structure.
 
-* `patterns_files_glob` (Optional): A glob `String` that describes which pattern files to use from the directories specified for `patterns_directories`. Default value is `*`
+```
+patterns_folder/
+  - patterns1.txt
+  - patterns2.txt
+extra_patterns_folder/
+  - extra_patterns1.txt
+```
 
+The following grok configuration will register all patterns in `patterns1.txt`, `patterns2.txt`, and `extra_patterns1.txt`
 
+```yaml
+prepper:
+  - grok:
+      patterns_directories: ["path/to/patterns_folder", "path/to/extra_patterns_folder"]
+      match:
+        message: ["%{CUSTOM_PATTERN_FROM_FILE:my_pattern}"]
+```
+
+When adding custom patterns to a file, one pattern should be declared per line. A space should separate the pattern name and its regex. The following example declares two custom patterns, `DOG` and `CAT`.
+
+```
+DOG beagle|chihuaha|retriever
+CAT persian|siamese|siberian
+```
+&nbsp;
+* `patterns_files_glob` (Optional): A glob `String` that describes which pattern files to use from the directories specified for `patterns_directories`. Default value is `*`<br></br>
+
+  
 * `target_key` (Optional): A `String` that will wrap all captures for a Record in an additional outer key value. Default value is `null`
 
 
-* `timeout_millis` (Optional): An `int` that specifies the maximum amount of time, in milliseconds, that matching will be performed on an individual Record before it times out and moves on to the next Record.
-Setting a `timeout_millis = 0` will make it so that matching a Record never times out. Default value is `30,000`
+  Given the same setup from [Basic Grok Example](#basic-grok-example), modify the `pipeline.yaml` grok configuration to add a `target_key` named `grokked`
 
-## Notes on Patterns
+```yaml
+  prepper:
+    - grok:
+        target_key: "grok"
+        match:
+          message: ['%{IPORHOST} \[%{HTTPDATE:timestamp}\] %{NUMBER:response_status:int}']
+```
 
-The Grok Prepper uses the [java-grok Library](https://github.com/thekrakken/java-grok) internally and supports all java-grok library compatible patterns. The java-grok library is built using the `java.util.regex` regular expression library.
+The resulting grokked log will now look like this.
+
+```
+{ 
+  "message":"127.0.0.1 198.126.12 [10/Oct/2000:13:55:36 -0700] 200",
+  "grokked": {
+     "response_status":200,
+     "clientip":"198.126.12",
+     "timestamp":"10/Oct/2000:13:55:36 -0700"
+  }
+}
+```
+
+All of the grok captures were wrapped in an outer key named `grokked`.<br></br>
 
-[Default Patterns](https://github.com/thekrakken/java-grok/blob/master/src/main/resources/patterns/patterns)
+* `timeout_millis` (Optional): An `int` that specifies the maximum amount of time, in milliseconds, that matching will be performed on an individual Record before it times out and moves on to the next Record.
```

Comment:
Shall we specify here whether the output record will remain the same as input if timeout happens?

Reply:
Good note!